src/cubic_loader/qlik/ods_qlik.py

```diff
@@ -0,0 +1,402 @@
+        current_schema = self.etl_status.last_schema
+        error_files: List[str] = []
+        new_column_set: Set[str] = set()
+        work_files = [(wf, self.etl_status) for wf in get_cdc_dfms(self.etl_status, self.table)]
+        with ThreadPoolExecutor(max_workers=threading_cpu_count()) as pool:
+            for result in pool.map(thread_load_cdc_file, work_files):
+                dfm_result, new_columns = result
+                if new_columns is None and dfm_result is not None:
+                    # load was success, save cdc_ts if needed
+                    current_cdc_ts = max(current_cdc_ts, dfm_result)
+                elif new_columns is not None and dfm_result:
+                    # handle file with new columns added
+                    error_files.append(dfm_result)
+                    for column in new_columns:
+                        new_column_set.add(json.dumps(column))
+
+        if len(error_files) > 0:
+            # handle new column additions
+            new_columns_to_add: List[DFMSchemaFields] = [json.loads(column) for column in new_column_set]
+            self.db.execute(add_columns_to_table(new_columns_to_add, self.db_fact_table))
+            for column in new_columns_to_add:
+                current_schema.append(column)
+            self.update_status(last_schema=current_schema)
+            # re-process failed files with added columns
+            work_files = [
+                (DFMDetails(path=path, ts=re_get_first(path, RE_CDC_TS)), self.etl_status) for path in error_files
+            ]
+            with ThreadPoolExecutor(max_workers=threading_cpu_count()) as pool:
+                for result in pool.map(thread_load_cdc_file, work_files):
+                    dfm_result, new_columns = result
+                    if new_columns is None and dfm_result is not None:
+                        current_cdc_ts = max(current_cdc_ts, dfm_result)
+
+        self.db.vaccuum_analyze(f"{ODS_SCHEMA}.{self.db_history_table}")
+
+        self.update_status(
+            last_cdc_ts=current_cdc_ts,
+            last_schema=current_schema,
+        )
+
+    def rds_fact_table_load(self) -> None:
+        """Load FACT Table records from History Table"""
+        logger = ProcessLogger("load_fact_table", table=self.db_fact_table)
+        schema = self.etl_status.last_schema
+
+        fact_table = f"{ODS_SCHEMA}.{self.db_fact_table}"
+        history_table = f"{ODS_SCHEMA}.{self.db_history_table}"
+
+        table_columns = [col["name"] for col in schema]
+        table_column_str = ",".join(table_columns)
+        key_columns = [col["name"] for col in schema if col["primaryKeyPos"] > 0]
+        key_str = ",".join(key_columns)
+
+        first_vals = []
+        for column in ["header__change_oper"] + table_columns:
+            if column in key_columns:
+                continue
+            q = f"(array_remove(array_agg({column} ORDER BY header__timestamp DESC), NULL))[1] as {column}"
```

Comment:
From testing this query is the most optimized way to aggregate each individual column to find that last non-null value that was indicated by a CDC update file.